lib/spack/spack/test/cmd/compiler.py

```diff
@@ -103,3 +105,92 @@ def test_compiler_add(
     new_compiler = new_compilers - old_compilers
     assert any(c.version == spack.version.Version(mock_compiler_version)
                for c in new_compiler)
+
+
+@pytest.fixture
+def clangdir(tmpdir):
+    """Create a directory with some dummy compiler scripts in it.
+
+    Scripts are:
+      - clang
+      - clang++
+      - gfortran-8
+
+    """
+    with tmpdir.as_cwd():
+        with open('clang', 'w') as f:
+            f.write("""\
+#!/bin/sh
+if [ "$1" = "--version" ]; then
+    echo "Apple clang version 11.0.0 (clang-1100.0.33.16)"
```

Comment:
@alalazo: this is ready for review.  It works in its current form, but can you take a look at one thing for me?

On my machine, if I change this line to:

```
     echo "clang version 11.0.0 (clang-1100.0.33.16)"
```
And if I change the spec assertions below to look like this:

```
     assert new['spec'] == "clang@11.0.0"
```

Then `test_compiler_find_mixed_suffixes` will succeed when run alone but fail when running like this:

```
spack test lib/spack/spack/test/cmd/compiler.py
```

I am not sure why.  It seems like if you run that way, `gfortran-8` is detected as `apple-clang` but not `clang`, and things are not properly grouped under `clang` compiler ids.

Is there some race in the code where `gfortran-8` can only be associated with *one* compiler?

Reply:
I think it's memoization getting in the way of tests:

https://github.com/spack/spack/blob/bab18523402d610a7cddd7e7c59c95fe4cfe5ff6/lib/spack/spack/compiler.py#L30-L31

Doing what you say, with memoization enabled I get a failure because a few functions return the results of the tests that were run previously. The chain of calls is usually:

-  `<SpecificCompiler>.default_version` -> `spack.compiler.get_compiler_version_output`

If I remove memoization the tests pass in both cases.